File: ai_dj/state.py
"""Deterministic DJ state.

Everything that does not need a model lives here: tempo, key, mode, energy,
the section plan, which layer to change next, and a deterministic variation
used between model calls. The LLM only ever produces one small layer patch.

State is the single source of truth for the live set. It renders the full
Sonic Pi script on demand and can parse a saved script back into layers so a
session can be resumed.
"""

import random
import re
import time

# (name, seconds, target_energy) — a deterministic energy arc that loops.
SECTIONS = [
    ("intro", 60, 0.30),
    ("build", 75, 0.55),
    ("peak", 90, 0.85),
    ("breakdown", 45, 0.25),
    ("peak", 90, 0.90),
    ("outro", 45, 0.35),
]

# Layer priorities for deterministic decisions.
BUILD_ORDER = ["hats", "perc", "bass", "pad", "arp", "lead", "fx", "sub", "drone"]
REDUCE_ORDER = ["lead", "arp", "fx", "perc", "hats", "drone", "sub", "pad", "bass"]
CORE_LAYERS = {"kick", "bass"}

# the single "layer" an opaque set is evolved as
WHOLE_LAYER = "set"


def _align_strudel_identity(code, key, mode):
    """Rewrite a Strudel layer so its tonality matches the session.

    The model is told to stay in key, but a seed or replace can still emit
    `.scale("a:minor")` in an F-minor session. Rewrite the root:mode inside
    `.scale(...)` and `.chord(...)` so the music cannot contradict the session.
    """
    def fix(m):
        return f'{m.group(1)}"{key}:{mode}"'

    code = re.sub(r'(\.scale\()"[^"]*"', fix, code)
    code = re.sub(r'(\.chord\()"[^"]*"', fix, code)
    return code


def parse_layers(script, lang="sonic_pi"):
    """Split a rendered script into {layer_name: layer_code}.

    sonic_pi: top-level `live_loop :name do` blocks (what render() guarantees).
    strudel:  top-level `$ name: <pattern>` marker comments that render() emits.
    """
    if lang == "strudel":
        return _parse_layers_strudel(script)
    matches = list(re.finditer(r"(?m)^live_loop\s+:(\w+)\s+do", script))
    layers = {}
    for i, m in enumerate(matches):
        start = m.start()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(script)
        layers[m.group(1)] = script[start:end].strip()
    return layers


def _parse_layers_strudel(script):
    """Split a Strudel script on `// layer: <name>` markers (indentation ok)."""
    matches = list(re.finditer(r"(?m)^\s*//\s*layer:\s*(\w+)\s*$", script))
    layers = {}
    for i, m in enumerate(matches):
        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(script)
        body = script[start:end].strip().rstrip(",").strip()
        if i + 1 == len(matches):
            # the last layer runs to the end of the script, which includes the
            # stack's closing paren — drop it (and any blank lines before it)
            body = re.sub(r"\n\s*\)\s*$", "", body).strip()
        if body:
            layers[m.group(1)] = body
    return layers


def parse_header(script):
    """Read the '# state ...' / '// state ...' comment line back into a dict."""
    m = re.search(r"(?m)^(?:\#|//)\s*state\s+(.+)$", script)
    if not m:
        return {}
    out = {}
    for tok in m.group(1).split():
        if "=" in tok:
            k, v = tok.split("=", 1)
            out[k] = v
    return out


class DJState:
    def __init__(self, bpm, key, mode, layers, energy=0.35, mood="",
                 section_index=0, model="", lang="sonic_pi", opaque=False):
        self.bpm = int(bpm)
        self.key = key
        self.mode = mode
        self.lang = lang
        self.layers = dict(layers)
        self.energy = float(energy)
        self.mood = mood
        self.section_index = section_index % len(SECTIONS)
        self.section_started = time.time()
        self.model = model
        self.last_action = ""
        self.last_layer = None
        self.history = []
        self.rng = random.Random(bpm * 1000 + len(layers))
        # An opaque set is a whole script the user pasted (or that we resumed)
        # which does not decompose into `// layer:` blocks. It is the source of
        # truth: rendered verbatim, and evolved as one piece — never rebuilt
        # from the layer model, which would resurrect the previous set.
        self.opaque = opaque

    # -- section / energy ---------------------------------------------------
    @property
    def section(self):
        return SECTIONS[self.section_index][0]

    def target_energy(self):
        return SECTIONS[self.section_index][2]

    def section_seconds(self):
        return SECTIONS[self.section_index][1]

    def section_elapsed(self):
        return time.time() - self.section_started

    def advance_section(self):
        self.section_index = (self.section_index + 1) % len(SECTIONS)
        self.section_started = time.time()

    # -- deterministic planning --------------------------------------------
    def _pick_existing(self, exclude_last=True):
        names = [n for n in self.layers if not (exclude_last and n == self.last_layer)]
        if not names:
            names = list(self.layers)
        return self.rng.choice(names) if names else None

    def plan_next(self):
        """Deterministically choose the layer and direction for the next step."""
        target = self.target_energy()
        delta = target - self.energy
        if self.opaque:
            # one piece: never add or drop "layers" it does not have
            if delta > 0.08:
                direction = "build"
            elif delta < -0.08:
                direction = "reduce"
            else:
                direction = "vary"
            return {"layer": WHOLE_LAYER, "direction": direction}
        if delta > 0.08:
            direction = "build"
            layer = next((l for l in BUILD_ORDER if l not in self.layers), None)
            if layer is None:
                direction = "intensify"
                layer = self._pick_existing()
        elif delta < -0.08:
            direction = "reduce"
            layer = next((l for l in REDUCE_ORDER
                          if l in self.layers and l not in CORE_LAYERS), None)
            if layer is None:
                direction = "vary"
                layer = self._pick_existing()
        else:
            direction = "vary"
            layer = self._pick_existing()
        return {
            "direction": direction,
            "layer": layer,
            "energy_now": round(self.energy, 2),
            "target_energy": round(target, 2),
            "section": self.section,
        }

    # -- applying a model patch --------------------------------------------
    def apply_op(self, decided):
        op = decided.get("op") or {}
        kind = op.get("kind", "modify")
        layer = op.get("layer") or self.last_layer
        ruby = (decided.get("ruby") or "").strip()
        if self.opaque:
            # the set is one piece: a patch can only replace it wholesale, and
            # a model that names some other "layer" must not add one (render()
            # would ignore it and the change would silently vanish)
            kind, layer = "modify", WHOLE_LAYER
        if kind == "remove" and layer:
            self.layers.pop(layer, None)
        elif layer and ruby:
            self.layers[layer] = ruby
        self.last_layer = layer
        action = (decided.get("decision") or {}).get("action", "")
        self.last_action = action
        self.history.append(action)
        self.history = self.history[-5:]
        self.mood = (decided.get("hearing") or {}).get("mood", self.mood)
        # step energy toward the section target
        if kind == "remove":
            self.energy = max(0.0, self.energy - 0.15)
        else:
            self.energy += 0.2 * (self.target_energy() - self.energy)
        self.energy = max(0.0, min(1.0, self.energy))

    # -- deterministic variation (no model) --------------------------------
    def variation(self):
        """Nudge one layer's numeric params. Safe no-op if nothing matches."""
        layer = self._pick_existing()
        if not layer:
            return False
        code = self.layers[layer]

        def bump(m):
            val = float(m.group(2))
            factor = self.rng.choice([0.9, 0.95, 1.05, 1.1])
            return f"{m.group(1)}{round(val * factor, 3)}"

        if self.lang == "strudel":
            # method-call params, e.g. .gain(0.3) .room(0.4) .lpf(2000)
            params = ("gain", "room", "lpf", "hpf", "attack", "release",
                      "delay", "pan", "shape")
        else:
            # Sonic Pi keyword args, e.g. amp: 0.5
            params = ("amp", "cutoff", "release")

        new = code
        for param in params:
            if self.lang == "strudel":
                pat = rf"(\.{param}\(\s*)([0-9.]+)"
            else:
                pat = rf"(\b{param}:\s*)([0-9.]+)"
            new = re.sub(pat, bump, code, count=1)
            if new != code:
                break
        if new != code:
            self.layers[layer] = new
            self.last_layer = layer
            return True
        return False

    # -- context + render ---------------------------------------------------
    def to_context(self):
        return (
            f"bpm={self.bpm} key={self.key} mode={self.mode} "
            f"energy={self.energy:.2f} section={self.section}\n"
            f"active layers: {', '.join(self.layers) or 'none'}\n"
            f"last action: {self.last_action or 'none'}")

    def render(self):
        if self.opaque:
            # the user's own script, played as-is
            return self.layers.get(WHOLE_LAYER, "")
        if self.lang == "strudel":
            return self._render_strudel()
        header = (
            "# ai-dj live - generated script\n"
            f"# model: {self.model}\n"
            f"# state bpm={self.bpm} key={self.key} mode={self.mode} "
            f"energy={self.energy:.2f} section={self.section}\n"
            f"# last: {self.last_action}\n"
            "\n")
        body = f"use_bpm {self.bpm}\n\n" + "\n\n".join(self.layers.values())
        return header + body

    def _render_strudel(self):
        header = (
            "// ai-dj live - generated script\n"
            f"// model: {self.model}\n"
            f"// state bpm={self.bpm} key={self.key} mode={self.mode} "
            f"energy={self.energy:.2f} section={self.section}\n"
            f"// last: {self.last_action}\n"
            f"\nsetcpm({self.bpm}/4)\n\n"
            "stack(\n")
        parts = []
        for name, code in self.layers.items():
            parts.append(f"  // layer: {name}\n  {code}")
        return header + ",\n".join(parts) + "\n)\n"

    def set_manual(self, script):
        """Adopt a script the user wrote.

        If it decomposes into `// layer:` blocks we keep evolving it layer by
        layer. Otherwise it is opaque: kept verbatim, and evolved as one piece.
        Either way the previous layer model is dropped — otherwise the next
        render would resurrect the old set.
        """
        layers = parse_layers(script, self.lang)
        if layers:
            self.layers = layers
            self.opaque = False
        else:
            self.layers = {WHOLE_LAYER: script}
            self.opaque = True
        self.last_layer = None

    @classmethod
    def from_script(cls, script, model="", lang="sonic_pi"):
        layers = parse_layers(script, lang)
        h = parse_header(script)
        return cls(
            bpm=int(h.get("bpm", 120)),
            key=h.get("key", "a"),
            mode=h.get("mode", "minor"),
            layers=layers if layers else ({WHOLE_LAYER: script} if script.strip() else {}),
            energy=float(h.get("energy", 0.35)),
            model=model,
            lang=lang,
            opaque=not layers and bool(script.strip()),
        )

    def adopt_seed(self, ruby, hearing=None, lang=None, keep_identity=False):
        """Replace layers with those parsed from a seed script.

        keep_identity: keep this session's tempo/key/mode and only take the
        seed's musical content. Used for prompt seeding and replace feedback,
        where the session (or the archetype it was cast from) already has an
        identity and the model should write music inside it, not rename it.
        """
        lang = lang or self.lang
        layers = parse_layers(ruby or "", lang)
        if layers and keep_identity and lang == "strudel":
            # the model may still emit another key; make the music agree with
            # the session so the set cannot contradict itself
            layers = {n: _align_strudel_identity(c, self.key, self.mode)
                      for n, c in layers.items()}
        if layers:
            self.layers = layers
        if not keep_identity:
            m = re.search(r"(?:use_bpm|setcpm\()\s*(\d+)", ruby or "")
            if m:
                self.bpm = int(m.group(1))
        if hearing:
            if not keep_identity:
                k = (hearing.get("key") or "")
                root = re.search(r"\b([a-gA-G])", k)
                if root:
                    self.key = root.group(1).lower()
                low = k.lower()
                if "minor" in low:
                    self.mode = "minor"
                elif "major" in low:
                    self.mode = "major"
            self.mood = hearing.get("mood", self.mood)
        self.last_action = "seed"

File: packages/dj/src/state.ts
/** Deterministic DJ state.
 *
 * Everything that does not need a model lives here: tempo, key, mode, energy,
 * the section plan, which layer to change next, and a deterministic variation
 * used between model calls. The LLM only ever produces one small layer patch.
 *
 * State is the single source of truth for the live set. It renders the full
 * script on demand and can parse a saved script back into layers so a session
 * can be resumed.
 *
 * Ported from ai_dj/state.py — same behaviour, idiomatic names.
 */

// (name, seconds, target_energy) — a deterministic energy arc that loops.
export const SECTIONS: Array<readonly [string, number, number]> = [
  ["intro", 60, 0.3],
  ["build", 75, 0.55],
  ["peak", 90, 0.85],
  ["breakdown", 45, 0.25],
  ["peak", 90, 0.9],
  ["outro", 45, 0.35],
];

// Layer priorities for deterministic decisions.
export const BUILD_ORDER = ["hats", "perc", "bass", "pad", "arp", "lead", "fx", "sub", "drone"];
export const REDUCE_ORDER = ["lead", "arp", "fx", "perc", "hats", "drone", "sub", "pad", "bass"];
export const CORE_LAYERS = new Set(["kick", "bass"]);

// The single "layer" an opaque set is evolved as.
export const WHOLE_LAYER = "set";

const VARIATION_FACTORS = [0.9, 0.95, 1.05, 1.1];
const STRUDEL_PARAMS = ["gain", "room", "lpf", "hpf", "attack", "release", "delay", "pan", "shape"];
const SONIC_PARAMS = ["amp", "cutoff", "release"];

/** Deterministic PRNG (mulberry32) — the tests never depend on its sequence,
 * only on the fact that choices vary; mirrors random.Random's role. */
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function alignStrudelIdentity(code: string, key: string, mode: string): string {
  // Rewrite a Strudel layer so its tonality matches the session: the model is
  // told to stay in key, but a seed can still emit another root.
  const fix = (_m: string, g1: string) => `${g1}"${key}:${mode}"`;
  return code
    .replace(/(\.scale\()"[^"]*"/g, fix)
    .replace(/(\.chord\()"[^"]*"/g, fix);
}

export function parseLayers(script: string, lang = "sonic_pi"): Record<string, string> {
  if (lang === "strudel") return parseLayersStrudel(script);
  const matches = [...script.matchAll(/^live_loop\s+:(\w+)\s+do/gm)];
  const layers: Record<string, string> = {};
  for (let idx = 0; idx < matches.length; idx++) {
    const m = matches[idx];
    const start = m.index ?? 0;
    const end = idx + 1 < matches.length ? (matches[idx + 1].index ?? script.length) : script.length;
    layers[m[1]] = script.slice(start, end).trim();
  }
  return layers;
}

function parseLayersStrudel(script: string): Record<string, string> {
  const matches = [...script.matchAll(/^\s*\/\/\s*layer:\s*(\w+)\s*$/gm)];
  const layers: Record<string, string> = {};
  for (let idx = 0; idx < matches.length; idx++) {
    const m = matches[idx];
    const start = (m.index ?? 0) + m[0].length;
    const end = idx + 1 < matches.length ? (matches[idx + 1].index ?? script.length) : script.length;
    let body = script.slice(start, end).trim().replace(/,$/, "").trim();
    if (idx + 1 === matches.length) {
      // the last layer runs to the end of the script, which includes the
      // stack's closing paren — drop it (and any blank lines before it)
      body = body.replace(/\n\s*\)\s*$/, "").trim();
    }
    if (body) layers[m[1]] = body;
  }
  return layers;
}

export function parseHeader(script: string): Record<string, string> {
  const m = script.match(/^(?:#|\/\/)\s*state\s+(.+)$/m);
  if (!m) return {};
  const out: Record<string, string> = {};
  for (const tok of m[1].split(/\s+/)) {
    const eq = tok.indexOf("=");
    if (eq >= 0) out[tok.slice(0, eq)] = tok.slice(eq + 1);
  }
  return out;
}

export interface DJStateOptions {
  bpm: number;
  key: string;
  mode: string;
  layers: Record<string, string>;
  energy?: number;
  mood?: string;
  sectionIndex?: number;
  model?: string;
  lang?: string;
  opaque?: boolean;
}

export interface Plan {
  direction: string;
  layer: string | null;
  energyNow: number;
  targetEnergy: number;
  section: string;
}

export interface OpDecision {
  op?: { kind?: string; layer?: string };
  ruby?: string;
  decision?: { action?: string };
  hearing?: { mood?: string };
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

/** Like Python's str(round(x, 3)) for the bumped values we produce. */
function formatBump(v: number): string {
  const r = Math.round(v * 1000) / 1000;
  return Number.isInteger(r) ? r.toFixed(1) : String(r);
}

export class DJState {
  bpm: number;
  key: string;
  mode: string;
  lang: string;
  layers: Record<string, string>;
  energy: number;
  mood: string;
  sectionIndex: number;
  sectionStarted: number;
  model: string;
  lastAction = "";
  lastLayer: string | null = null;
  history: string[] = [];
  opaque: boolean;
  private rng: () => number;

  constructor(opts: DJStateOptions) {
    this.bpm = Math.trunc(opts.bpm);
    this.key = opts.key;
    this.mode = opts.mode;
    this.lang = opts.lang ?? "sonic_pi";
    this.layers = { ...opts.layers };
    this.energy = opts.energy ?? 0.35;
    this.mood = opts.mood ?? "";
    this.sectionIndex = (opts.sectionIndex ?? 0) % SECTIONS.length;
    this.sectionStarted = Date.now();
    this.model = opts.model ?? "";
    this.opaque = opts.opaque ?? false;
    this.rng = mulberry32(this.bpm * 1000 + Object.keys(this.layers).length);
  }

  get section(): string {
    return SECTIONS[this.sectionIndex][0];
  }

  targetEnergy(): number {
    return SECTIONS[this.sectionIndex][2];
  }

  sectionSeconds(): number {
    return SECTIONS[this.sectionIndex][1];
  }

  sectionElapsed(): number {
    return (Date.now() - this.sectionStarted) / 1000;
  }

  advanceSection(): void {
    this.sectionIndex = (this.sectionIndex + 1) % SECTIONS.length;
    this.sectionStarted = Date.now();
  }

  private pickExisting(excludeLast = true): string | null {
    let names = Object.keys(this.layers).filter(
      (n) => !(excludeLast && n === this.lastLayer),
    );
    if (!names.length) names = Object.keys(this.layers);
    if (!names.length) return null;
    return names[Math.floor(this.rng() * names.length)];
  }

  planNext(): Plan {
    const target = this.targetEnergy();
    const delta = target - this.energy;
    if (this.opaque) {
      // one piece: never add or drop "layers" it does not have
      const direction = delta > 0.08 ? "build" : delta < -0.08 ? "reduce" : "vary";
      return {
        direction, layer: WHOLE_LAYER,
        energyNow: round2(this.energy), targetEnergy: round2(target),
        section: this.section,
      };
    }
    let direction: string;
    let layer: string | null;
    if (delta > 0.08) {
      direction = "build";
      const found = BUILD_ORDER.find((l) => !(l in this.layers));
      layer = found ?? null;
      if (layer === null) {
        direction = "intensify";
        layer = this.pickExisting();
      }
    } else if (delta < -0.08) {
      direction = "reduce";
      const found = REDUCE_ORDER.find(
        (l) => l in this.layers && !CORE_LAYERS.has(l));
      layer = found ?? null;
      if (layer === null) {
        direction = "vary";
        layer = this.pickExisting();
      }
    } else {
      direction = "vary";
      layer = this.pickExisting();
    }
    return {
      direction, layer,
      energyNow: round2(this.energy), targetEnergy: round2(target),
      section: this.section,
    };
  }

  applyOp(decided: OpDecision): void {
    const op = decided.op ?? {};
    let kind = op.kind ?? "modify";
    let layer = op.layer ?? this.lastLayer;
    const ruby = (decided.ruby ?? "").trim();
    if (this.opaque) {
      // the set is one piece: a patch can only replace it wholesale, and
      // a model that names some other "layer" must not add one (render()
      // would ignore it and the change would silently vanish)
      kind = "modify";
      layer = WHOLE_LAYER;
    }
    if (kind === "remove" && layer) {
      delete this.layers[layer];
    } else if (layer && ruby) {
      this.layers[layer] = ruby;
    }
    this.lastLayer = layer;
    const action = decided.decision?.action ?? "";
    this.lastAction = action;
    this.history.push(action);
    this.history = this.history.slice(-5);
    this.mood = decided.hearing?.mood ?? this.mood;
    // step energy toward the section target
    if (kind === "remove") {
      this.energy = Math.max(0, this.energy - 0.15);
    } else {
      this.energy += 0.2 * (this.targetEnergy() - this.energy);
    }
    this.energy = Math.max(0, Math.min(1, this.energy));
  }

  /** Nudge one layer's numeric params. Safe no-op if nothing matches. */
  variation(): boolean {
    const layer = this.pickExisting();
    if (!layer) return false;
    const code = this.layers[layer];

    const bump = (m: string, g1: string, g2: string) => {
      const val = parseFloat(g2);
      const factor = VARIATION_FACTORS[Math.floor(this.rng() * VARIATION_FACTORS.length)];
      return `${g1}${formatBump(val * factor)}`;
    };

    const params = this.lang === "strudel"
      ? STRUDEL_PARAMS // method-call params, e.g. .gain(0.3) .room(0.4)
      : SONIC_PARAMS; // Sonic Pi keyword args, e.g. amp: 0.5

    let next = code;
    for (const param of params) {
      const pat = this.lang === "strudel"
        ? new RegExp(`(\\.${param}\\(\\s*)([0-9.]+)`)
        : new RegExp(`(\\b${param}:\\s*)([0-9.]+)`);
      next = code.replace(pat, bump);
      if (next !== code) break;
    }
    if (next !== code) {
      this.layers[layer] = next;
      this.lastLayer = layer;
      return true;
    }
    return false;
  }

  toContext(): string {
    return (
      `bpm=${this.bpm} key=${this.key} mode=${this.mode} ` +
      `energy=${this.energy.toFixed(2)} section=${this.section}\n` +
      `active layers: ${Object.keys(this.layers).join(", ") || "none"}\n` +
      `last action: ${this.lastAction || "none"}`
    );
  }

  render(): string {
    if (this.opaque) {
      // the user's own script, played as-is
      return this.layers[WHOLE_LAYER] ?? "";
    }
    if (this.lang === "strudel") return this.renderStrudel();
    const header =
      "# ai-dj live - generated script\n" +
      `# model: ${this.model}\n` +
      `# state bpm=${this.bpm} key=${this.key} mode=${this.mode} ` +
      `energy=${this.energy.toFixed(2)} section=${this.section}\n` +
      `# last: ${this.lastAction}\n` +
      "\n";
    return header + `use_bpm ${this.bpm}\n\n` + Object.values(this.layers).join("\n\n");
  }

  private renderStrudel(): string {
    const header =
      "// ai-dj live - generated script\n" +
      `// model: ${this.model}\n` +
      `// state bpm=${this.bpm} key=${this.key} mode=${this.mode} ` +
      `energy=${this.energy.toFixed(2)} section=${this.section}\n` +
      `// last: ${this.lastAction}\n` +
      `\nsetcpm(${this.bpm}/4)\n\n` +
      "stack(\n";
    const parts = Object.entries(this.layers).map(
      ([name, code]) => `  // layer: ${name}\n  ${code}`,
    );
    return header + parts.join(",\n") + "\n)\n";
  }

  setManual(script: string): void {
    // Adopt a script the user wrote. If it decomposes into `// layer:` blocks
    // we keep evolving it layer by layer. Otherwise it is opaque: kept
    // verbatim, and evolved as one piece. Either way the previous layer model
    // is dropped — otherwise the next render would resurrect the old set.
    const layers = parseLayers(script, this.lang);
    if (Object.keys(layers).length) {
      this.layers = layers;
      this.opaque = false;
    } else {
      this.layers = { [WHOLE_LAYER]: script };
      this.opaque = true;
    }
    this.lastLayer = null;
  }

  static fromScript(script: string, model = "", lang = "sonic_pi"): DJState {
    const layers = parseLayers(script, lang);
    const h = parseHeader(script);
    const hasLayers = Object.keys(layers).length > 0;
    return new DJState({
      bpm: h.bpm !== undefined ? parseInt(h.bpm, 10) : 120,
      key: h.key ?? "a",
      mode: h.mode ?? "minor",
      layers: hasLayers ? layers : script.trim() ? { [WHOLE_LAYER]: script } : {},
      energy: h.energy !== undefined ? parseFloat(h.energy) : 0.35,
      model,
      lang,
      opaque: !hasLayers && script.trim().length > 0,
    });
  }

  adoptSeed(ruby: string, hearing?: { key?: string; mood?: string } | null,
    lang?: string, keepIdentity = false): void {
    // Replace layers with those parsed from a seed script. With keepIdentity
    // the session keeps its tempo/key/mode and only takes the seed's musical
    // content — used for prompt seeding and replace feedback.
    const l = lang ?? this.lang;
    let layers = parseLayers(ruby ?? "", l);
    if (Object.keys(layers).length && keepIdentity && l === "strudel") {
      // the model may still emit another key; make the music agree with
      // the session so the set cannot contradict itself
      const aligned: Record<string, string> = {};
      for (const [n, c] of Object.entries(layers)) {
        aligned[n] = alignStrudelIdentity(c, this.key, this.mode);
      }
      layers = aligned;
    }
    if (Object.keys(layers).length) this.layers = layers;
    if (!keepIdentity) {
      const m = (ruby ?? "").match(/(?:use_bpm|setcpm\()\s*(\d+)/);
      if (m) this.bpm = parseInt(m[1], 10);
    }
    if (hearing) {
      if (!keepIdentity) {
        const root = (hearing.key ?? "").match(/\b([a-gA-G])/);
        if (root) this.key = root[1].toLowerCase();
        const low = (hearing.key ?? "").toLowerCase();
        if (low.includes("minor")) this.mode = "minor";
        else if (low.includes("major")) this.mode = "major";
      }
      this.mood = hearing.mood ?? this.mood;
    }
    this.lastAction = "seed";
  }
}
